src/ledger/shred_inserter/merkle_root_checks.zig

```diff
@@ -228,43 +197,88 @@ pub fn checkBackwardsChainedMerkleRootConsistency(
         .index = @intCast(prev_erasure_meta.first_received_code_index),
         .shred_type = .code,
     };
-    const prev_shred =
-        if (try shred_store.get(prev_shred_id)) |ps| ps else {
-        logger.warn().logf(&newlinesToSpaces(
-            \\Shred {any} indicated by the erasure meta {any} \
-            \\is missing from blockstore. This can happen if you have recently upgraded \
-            \\from a version < v1.18.13, or if blockstore cleanup has caught up to the root. \
-            \\Skipping the backwards chained merkle root consistency check
-        ), .{ prev_shred_id, prev_erasure_meta });
+
+    return checkAndReportMerkleRootConsistency(
+        .backward,
+        logger,
+        db,
+        shred_store,
+        duplicate_shreds,
+        shred,
+        prev_shred_id,
+    );
+}
+
+/// The input shreds must be from adjacent erasure sets in the same slot,
+/// or this function will not work correctly.
+fn checkAndReportMerkleRootConsistency(
+    direction: enum { forward, backward },
+    logger: Logger,
+    db: *BlockstoreDB,
+    shred_store: WorkingShredStore,
+    duplicate_shreds: *std.ArrayList(PossibleDuplicateShred),
+    shred: Shred,
+    other_shred_id: ShredId,
+) !bool {
+    const other_shred = if (try shred_store.get(other_shred_id)) |other_shred|
+        other_shred
+    else {
+        logger.warn().logf(
+            "Shred {any} is missing from blockstore. " ++
+                "This can happen if blockstore cleanup has caught up to the root. " ++
+                "Skipping the {} chained merkle root consistency check.",
+            .{ other_shred_id, direction },
+        );
         return true;
     };
-    const merkle_root = shred_mod.layout.getChainedMerkleRoot(prev_shred);
-    const chained_merkle_root = shred.chainedMerkleRoot() catch null;
 
-    if (!checkChaining(merkle_root, chained_merkle_root)) {
-        logger.warn().logf(&newlinesToSpaces(
-            \\Received conflicting chained merkle roots for slot: {}, shred {any} type {any} \
-            \\chains to merkle root {any}, however previous fec set code \
-            \\shred {any} has merkle root {any}. Reporting as duplicate
-        ), .{
+    const early_shred, const late_shred = switch (direction) {
```

Comment:
nit: kinda find `prev_shred/next_shred` easier to parse than `early_shred/late_shred`.

Reply:
I intentionally made this change to move away from "previous" and "next" because I felt that those names were misleading. When they're mixed up together in the same context, I feel that it implies they come before and after some central shred, implying there are three. But there are always only two shreds.

I get that early and late aren't the best names either. What about old_shred and new_shred?

Reply:
Gotcha. I think `old_shred` and `new_shred` is better. Or maybe `older_shred`/`newer_shred` to convey it is some sort of comparison.

Reply:
e827130473ba05e0cd0b88c3e4b6c6080f660f02